### PR TITLE
Correct fix of long comment

### LINKED ISSUE #923: Move inline comments in long lines before the calling code
Consider following line
```kotlin
val accessors = node.findAllDescendantsWithSpecificType(PROPERTY_ACCESSOR).filter { it.hasChildOfType(BLOCK) }  // exclude get with expression body
```
In short lines usage of inline comments quite often is convenient, so we don't need to move such comments at any time.
However, in long lines it's reduces the readability. 
Also in this rule we can use other definition of long lines, and move
comments not when line > 180 symbols, but earlier.

```kotlin
// exclude get with expression body
val accessors = node.findAllDescendantsWithSpecificType(PROPERTY_ACCESSOR).filter { it.hasChildOfType(BLOCK) }
```

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
* [x] Changed logic of the fix long comment
* [x] Updated test
* [x] Closes #923

### CODE REVIEW

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt [hunk 1]
```diff
@@ -237,13 +240,24 @@ class LineLength(configRules: List<RulesConfig>) : DiktatRule(
...
     private fun fixComment(wrongComment: LongLineFixableCases.Comment) {
         val wrongNode = wrongComment.node
-        val indexLastSpace = wrongComment.indexLastSpace
-        val nodeText = "//${wrongNode.text.substring(indexLastSpace, wrongNode.text.length)}"
-        wrongNode.treeParent.run {
-            addChild(LeafPsiElement(EOL_COMMENT, wrongNode.text.substring(0, indexLastSpace)), wrongNode)
-            addChild(PsiWhiteSpaceImpl("\n"), wrongNode)
-            addChild(LeafPsiElement(EOL_COMMENT, nodeText), wrongNode)
-            removeChild(wrongNode)
+        if (wrongComment.isNewLine) {
```
Comment (kgevorkyan) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt:243]: I suggest to add few comments in code about this logic - that we recognize two types of comments, which should be handled differently:
1) Ordinary comments, long part of which should be moved to the next line
2) Inline comments, which should be moved entirely to the previous line

File: diktat-rules/src/test/resources/test/paragraph3/long_line/LongLineCommentExpected.kt [hunk 2]
```diff
@@ -7,8 +7,8 @@ fun foo() {
     namesList.forEach { name ->
         if (name == "Nick") {
             namesList.map {
-                it.subSequence(0, 1) // This is
-// another comment inside map
+                // This is another comment inside map
```
Comment (kgevorkyan) [inline diktat-rules/src/test/resources/test/paragraph3/long_line/LongLineCommentExpected.kt:10]: Could you please provide more tests?
E.g.:
1) some comment, which shouldn't be moved as it too short and can be incline
```kotlin
val a = 5 // it's okay
```
2) some long comment, which should be moved before code, but still remains long and should be splited if possible
Before:
```kotlin
val value = "some string" // long long long comment
```
After, expected
```kotlin
// long long 
// long comment
val value = "some string"
```
Will we handle this situation?

**Comment by kgevorkyan** [comment]
Also I recommend to merge this PR after https://github.com/cqfn/diKTat/pull/932, since #932 affects to the common logic of this rule

File: diktat-rules/src/test/resources/test/paragraph3/long_line/LongLineCommentExpected.kt (hunk 2 above)
Comment (kentr0w) [reply diktat-rules/src/test/resources/test/paragraph3/long_line/LongLineCommentExpected.kt:10]: No. I thought logic should be as in rule. If we found long line and after fix this line still long we don't fix it again.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt (hunk 1 above)
Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt:243]: Agree, added

File: diktat-rules/src/test/resources/test/paragraph3/long_line/LongLineCommentExpected.kt (hunk 2 above)
Comment (kentr0w) [reply diktat-rules/src/test/resources/test/paragraph3/long_line/LongLineCommentExpected.kt:10]: We have same logic with `BinaryExpression`

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt [hunk 3]
```diff
@@ -237,13 +240,24 @@ class LineLength(configRules: List<RulesConfig>) : DiktatRule(
...
+            val indexLastSpace = wrongComment.indexLastSpace
+            val nodeText = "//${wrongNode.text.substring(indexLastSpace, wrongNode.text.length)}"
+            wrongNode.treeParent.run {
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt:246]:
```suggestion
            wrongNode.treeParent.apply {
```
`apply` indicates, that returned value is not needed and the scope is only intended to use the receiver object

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt [hunk 4]
```diff
@@ -237,13 +240,24 @@ class LineLength(configRules: List<RulesConfig>) : DiktatRule(
...
+                removeChild(wrongNode)
+            }
+        } else {
+            val grandParent = wrongNode.treeParent.treeParent
+            val parent = wrongNode.treeParent
+            if (wrongNode.treePrev.isWhiteSpace()) {
+                parent.removeChild(wrongNode.treePrev)
+            }
+            parent.removeChild(wrongNode)
+            grandParent?.addChild(wrongNode, parent)
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt:259]: What if `grandParent` is `null`? Then we simply remove the comment and do nothing?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt [hunk 5]
```diff
@@ -421,7 +435,16 @@ class LineLength(configRules: List<RulesConfig>) : DiktatRule(
...
-        class Comment(val node: ASTNode, val indexLastSpace: Int) : LongLineFixableCases()
+        /**
+         * @param node node
+         * @param isNewLine  flag to handle type of comment: ordinary comment(long part of which should be moved to the next line)
+         * and inline comments (which should be moved entirely to the previous line)
+         * @param indexLastSpace index of last space to substring comment
+         */
+        class Comment(
+            val node: ASTNode,
+            val isNewLine: Boolean,
```
Comment (petertrr) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt:446]: `hasNewLineBefore` or something like that would be more clear, I think

**Review by petertrr — CHANGES_REQUESTED** [review]

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt [hunk 6]
```diff
@@ -41,6 +41,8 @@ import com.pinterest.ktlint.core.ast.ElementType.RPAR
 import com.pinterest.ktlint.core.ast.ElementType.SHORT_STRING_TEMPLATE_ENTRY
 import com.pinterest.ktlint.core.ast.ElementType.STRING_TEMPLATE
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt:42]: could you also update Diktat code style document with this rules also?

Comment (kentr0w) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/LineLength.kt:42]: Added

**Review by akuleshov7 — APPROVED** [review]

**Review by petertrr — APPROVED** [review]

**Pull request merged**